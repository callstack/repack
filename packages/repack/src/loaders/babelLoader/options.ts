import type { LoaderContext } from '@rspack/core';
import type { validate } from 'schema-utils';

export interface BabelLoaderOptions {
  projectRoot: string;
}

type Schema = Parameters<typeof validate>[0];

export const optionsSchema: Schema = {
  type: 'object',
  required: ['projectRoot'],
  properties: {
    projectRoot: { type: 'string' },
  },
};

export function getOptions(
  loaderContext: LoaderContext<BabelLoaderOptions>
): BabelLoaderOptions {
  const options = loaderContext.getOptions(loaderContext) || {};

  //   validate(optionsSchema, options, { name: 'repackBabelLoader' });

  return options;
}
