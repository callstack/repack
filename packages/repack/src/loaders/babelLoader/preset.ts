import type { TransformOptions } from '@babel/core';

export const repackBabelPreset: TransformOptions = {
  plugins: [
    ['@babel/plugin-syntax-typescript', false],
    ['@react-native/babel-plugin-codegen', false],
    '@babel/plugin-transform-flow-strip-types',
  ],
  overrides: [
    {
      test: /(?:^|[\\/])(?:Native\w+|(\w+)NativeComponent)\.[jt]sx?$/,
      plugins: ['@react-native/babel-plugin-codegen'],
    },
    {
      test: /\.ts$/,
      plugins: [
        [
          '@babel/plugin-syntax-typescript',
          { isTSX: false, allowNamespaces: true },
        ],
      ],
    },
    {
      test: /\.tsx$/,
      plugins: [
        [
          '@babel/plugin-syntax-typescript',
          { isTSX: true, allowNamespaces: true },
        ],
      ],
    },
  ],
};
