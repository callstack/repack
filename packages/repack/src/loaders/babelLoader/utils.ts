import type { ParseResult } from '@babel/core';

interface HermesParser {
  parse: (
    src: string,
    opts: {
      babel: boolean;
      flow?: 'all' | 'detect';
      reactRuntimeTarget: string;
      sourceType: 'script' | 'module' | 'unambiguous' | null | undefined;
    }
  ) => ParseResult;
}

export function isTypeScriptSource(fileName: string) {
  return !!fileName && fileName.endsWith('.ts');
}

export function isTSXSource(fileName: string) {
  return !!fileName && fileName.endsWith('.tsx');
}

function resolveHermesParser(projectRoot: string) {
  const reactNativeBabelPresetPath = require.resolve(
    '@react-native/babel-preset',
    { paths: [projectRoot] }
  );

  const babelPluginSyntaxHermesParserPath = require.resolve(
    'babel-plugin-syntax-hermes-parser',
    { paths: [reactNativeBabelPresetPath] }
  );

  const hermesParserPath = require.resolve('hermes-parser', {
    paths: [babelPluginSyntaxHermesParserPath],
  });

  return hermesParserPath;
}

export async function loadHermesParser(
  projectRoot?: string | null,
  providedHermesParserPath?: string
): Promise<HermesParser> {
  try {
    const hermesParserPath =
      providedHermesParserPath ??
      resolveHermesParser(projectRoot ?? process.cwd());
    const hermesParser = await import(hermesParserPath);
    return hermesParser;
  } catch (e) {
    console.error(e);
    throw new Error(
      `Failed to import 'hermes-parser'. Make sure you have '@react-native/babel-preset' installed in your project.`
    );
  }
}

const IGNORED_REPACK_FILENAMES = [
  'IncludeModules.js',
  'WebpackHMRClient.js',
].map((name) => name.replace(/\./g, '\\.'));

const IGNORED_REPACK_PATHS_REGEX = new RegExp(
  `repack/dist/modules/(${IGNORED_REPACK_FILENAMES.join('|')})$`
);

export function isIgnoredRepackDeepImport(filename: string): boolean {
  return IGNORED_REPACK_PATHS_REGEX.test(filename);
}
