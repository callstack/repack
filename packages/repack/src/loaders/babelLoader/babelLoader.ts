import fs from 'node:fs';
import path from 'node:path';
import {
  type Node,
  type TransformOptions,
  loadOptions,
  parseSync,
  transformFromAstSync,
} from '@babel/core';
// @ts-ignore
import * as hermesParser from 'hermes-parser';

import type { LoaderContext } from '@rspack/core';
import { type BabelLoaderOptions, getOptions } from './options.js';

export const raw = false;

function isTypeScriptSource(fileName: string) {
  return !!fileName && fileName.endsWith('.ts');
}

function isTSXSource(fileName: string) {
  return !!fileName && fileName.endsWith('.tsx');
}

interface CustomOptions {
  enableBabelRCLookup?: boolean;
  extendsBabelConfigPath?: string;
  excludePlugins?: string[];
  projectRoot: string;
}
/**
 * Return a memoized function that checks for the existence of a
 * project level .babelrc file, and if it doesn't exist, reads the
 * default RN babelrc file and uses that.
 */
const _getBabelRC = (() => {
  let babelRC: TransformOptions | null = null;

  return function _getBabelRC({
    projectRoot,
    extendsBabelConfigPath,
  }: CustomOptions) {
    if (babelRC != null) {
      return babelRC;
    }

    babelRC = {
      plugins: [],
      extends: extendsBabelConfigPath,
    };

    if (extendsBabelConfigPath) {
      return babelRC;
    }

    // Let's look for a babel config file in the project root.
    let projectBabelRCPath: string | undefined;

    // .babelrc
    if (projectRoot) {
      projectBabelRCPath = path.resolve(projectRoot, '.babelrc');
    }

    if (projectBabelRCPath) {
      // .babelrc.js
      if (!fs.existsSync(projectBabelRCPath)) {
        projectBabelRCPath = path.resolve(projectRoot, '.babelrc.js');
      }

      // babel.config.js
      if (!fs.existsSync(projectBabelRCPath)) {
        projectBabelRCPath = path.resolve(projectRoot, 'babel.config.js');
      }

      // If we found a babel config file, extend our config off of it
      if (fs.existsSync(projectBabelRCPath)) {
        babelRC.extends = projectBabelRCPath;
      }
    }

    return babelRC;
  };
})();

function buildBabelConfig(
  filename: string,
  options: CustomOptions
): TransformOptions {
  const extraConfig: TransformOptions = {
    babelrc: options.enableBabelRCLookup ?? true,
    code: true,
    cwd: options.projectRoot,
    filename,
    highlightCode: true,
    compact: false,
    comments: true,
    minified: false,
    plugins: [],
  };

  if (isTypeScriptSource(filename)) {
    extraConfig.plugins!.push([
      '@babel/plugin-syntax-typescript',
      { isTSX: false, allowNamespaces: true },
    ]);
  }

  if (isTSXSource(filename)) {
    extraConfig.plugins!.push([
      '@babel/plugin-syntax-typescript',
      { isTSX: true, allowNamespaces: true },
    ]);
  }

  const babelConfig = loadOptions(extraConfig) as any;

  if (options.excludePlugins) {
    babelConfig.plugins = babelConfig.plugins.filter(
      (plugin: { key: string }) => {
        return !options.excludePlugins!.includes(plugin.key);
      }
    );
  }

  // babelConfig.plugins.forEach((plugin: any) => {
  //   console.log(plugin.key);
  // });

  return babelConfig;
}

const transform = ({
  filename,
  options,
  src,
}: {
  filename: string;
  options: CustomOptions;
  src: string;
}) => {
  const builtConfig = buildBabelConfig(filename, options);
  const babelConfig: TransformOptions = {
    sourceType: 'unambiguous',
    ...builtConfig,
    caller: { name: 'repack' },
    ast: false,
    cloneInputAst: false,
  };

  const sourceAst: Node =
    isTypeScriptSource(filename) || isTSXSource(filename)
      ? parseSync(src, babelConfig)
      : hermesParser.parse(src, {
          babel: true,
          flow: 'all',
          reactRuntimeTarget: '19',
          sourceType: babelConfig.sourceType,
        });

  return transformFromAstSync(sourceAst, src, babelConfig);
};

export default function babelLoader(
  this: LoaderContext<BabelLoaderOptions>,
  source: string
) {
  this.cacheable();
  const callback = this.async();
  const options = getOptions(this);

  try {
    const result = transform({
      filename: this.resourcePath,
      src: source,
      options: {
        enableBabelRCLookup: true,
        excludePlugins: options.excludePlugins,
        // this is currently broken in Rspack and needs to be fixed upstream
        // for now we can pass this as an option to loader
        projectRoot: options.projectRoot,
      },
    });
    // @ts-ignore
    callback(null, result.code, result.map);
  } catch (e) {
    callback(e as Error);
  }
}
