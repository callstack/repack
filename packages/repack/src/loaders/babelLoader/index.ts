import babelLoader, { raw } from './babelLoader.js';

export { raw };
export default babelLoader;
