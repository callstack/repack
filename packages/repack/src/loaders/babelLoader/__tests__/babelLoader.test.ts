import { transform } from '../babelLoader.js';

jest.mock('../utils.js', () => {
  const actual = jest.requireActual('../utils.js');
  const { parseSync } = require('@babel/core');
  return {
    ...actual,
    loadHermesParser: jest.fn(async () => ({
      parse: (
        src: string,
        opts: { sourceType?: 'script' | 'module' | 'unambiguous' }
      ) =>
        parseSync(src, {
          sourceType: opts?.sourceType ?? 'unambiguous',
        }),
    })),
  };
});

describe('babelLoader', () => {
  describe('includePlugins', () => {
    it('includes @babel/plugin-transform-react-jsx and transforms JSX', async () => {
      const src = 'export const Component = () => <View test={1} />;';

      const result = await transform(
        src,
        {
          caller: { name: 'jest-babel-loader-test' },
          filename: '/virtual/Component.tsx',
          sourceMaps: false,
          sourceFileName: '/virtual/Component.tsx',
          sourceRoot: '/virtual',
          envName: 'production',
        },
        {
          includePlugins: [
            [
              '@babel/plugin-transform-react-jsx',
              { runtime: 'automatic', importSource: 'react' },
            ],
          ],
        }
      );

      expect(result.code).toMatchSnapshot();
    });
  });

  describe('excludePlugins', () => {
    const esmSrc = 'import foo from "bar"; export default foo;';

    it('by default transforms ESM modules to CJS (baseline)', async () => {
      const result = await transform(esmSrc, {
        caller: { name: 'jest-babel-loader-test' },
        filename: '/virtual/esm.ts',
        sourceMaps: false,
        sourceFileName: '/virtual/esm.ts',
        sourceRoot: '/virtual',
        envName: 'production',
      });

      expect(result.code).toMatchSnapshot();
    });

    it('excludes transform-modules-commonjs so ESM stays intact', async () => {
      const result = await transform(
        esmSrc,
        {
          caller: { name: 'jest-babel-loader-test' },
          filename: '/virtual/esm.ts',
          sourceMaps: false,
          sourceFileName: '/virtual/esm.ts',
          sourceRoot: '/virtual',
          envName: 'production',
        },
        { excludePlugins: ['transform-modules-commonjs'] }
      );

      expect(result.code).toMatchSnapshot();
    });
  });
});
