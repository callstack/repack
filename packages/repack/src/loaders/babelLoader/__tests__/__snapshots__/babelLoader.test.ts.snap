// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babelLoader excludePlugins by default transforms ESM modules to CJS (baseline) 1`] = `
""use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _bar = _interopRequireDefault(require("bar"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
var _default = exports.default = _bar.default;"
`;

exports[`babelLoader excludePlugins excludes transform-modules-commonjs so ESM stays intact 1`] = `
"import foo from "bar";
export default foo;"
`;

exports[`babelLoader includePlugins includes @babel/plugin-transform-react-jsx and transforms JSX 1`] = `
""use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Component = void 0;
var _jsxRuntime = require("react/jsx-runtime");
const Component = () => /*#__PURE__*/(0, _jsxRuntime.jsx)(View, {
  test: 1
});
exports.Component = Component;"
`;
